### ISSUE TITLE
Index out of range in PostList

### ISSUE THREAD

**Issue by guidedways** [aciodev/Stellar#9]
Crashed on line 304, `childData[row]` out of range:

```swift
        let pvc = PostListCell(postViewer: postViewer, childData: self.childData[row], votingColorProfile: votingColorProfile, mediaHandler: mediaHandler) // HERE: Index out of range
```

**Comment by guidedways** [comment]
A quick glance shows that this might be due to this line here:

`self.childData.append(contentsOf: redditPosts)` (line 183) in `PostList.swift` and `clearTempData()`

It's asynchronously changing the contents of the array while the user may separately try and scroll the table. It should use a temporary array to load and only switch / assign the new results to `childData` on the main thread to avoid this.

**Comment by aciodev** [comment]
This one might be a semantic issue I am misunderstanding. 

Potential source would be when the collection is [updated](https://github.com/aciodev/Stellar/blob/main/Stellar/UI/PostList/PostList.swift#L187). 

Taking a peak into the concurrent modification brings us to [this method](https://github.com/aciodev/Stellar/blob/main/Stellar/Reddit/RedditRequests.swift#L374-L422). And, more interestingly, [this line](https://github.com/aciodev/Stellar/blob/main/Stellar/Reddit/RedditRequests.swift#L419-L421).

My understanding is that ```DispatchQueue.main.async``` is executed on the main thread.  [This](https://www.hackingwithswift.com/read/9/4/back-to-the-main-thread-dispatchqueuemain) seems to support my claim. 

Were you just scrolling the PostList? Or did you perhaps trigger an update by switching views to a specific subreddit/timeline?

Will take a look into this. My first thought is that it might not be a concurrent modification. Will update when I can reproduce.

**Comment by aciodev** [comment]
Curious if it might be [this](https://github.com/aciodev/Stellar/blob/main/Stellar/UI/ViewController.swift#L123-L130). Cannot presently recall whether or not the observers are called on the UI thread or some background thread.

**Comment by guidedways** [comment]
That's right, I didn't drill into the method, it seems this does indeed get called on the main thread. However, it seems the `childData` array gets cleared here asynchronously: `clearTempData()`

So this is what's happening:

1) PostList seems to load and then `reloaData` gets called
2) Asynchronously `clearTempData()` gets called (the token refreshing in the background) but `tableView.reloadData()` is not - at this point `childData` is empty, but the table doesn't know this, so scrolling / clicking on the tableView causes a crash.

`reloadData()` must either be called immediately after `childData` is cleared, or childData must only be cleared at the same time reloadData is about to be called. It's okay to keep `childData` around the way it is (with stale data) until we're ready to fill it with new data.

I would propose the following change:

1) Delete the line `childData.removeAll()` from PostList.swift in `clearTempData()`
2) And insert this before `self.childData.append(contentsOf: redditPosts)` like so:

```swift
...
      self.subredditID = subredditID
      self.nextPageKey = pageKey
      
      self.childData.removeAll()
      self.childData.append(contentsOf: redditPosts)
      
...
```

This should fix all the index out of range crashes.

**Comment by aciodev** [comment]
This proposal would discard entries upon scroll.

Suppose batch sizes are N = 10. You initially load N. When the next N are loaded, the previous are discarded completely. 

You can verify this happens in the PostList by:
1. Remembering the top post in the PostList
2. Scroll down until you trigger a refresh
3. Scroll back to the top
